packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -509,28 +512,51 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
         // only if it is first failure, otherwise fallback to get call.
         if (usePost) {
             try {
-                return this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, true);
+                const odspSnapshot = await this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, true);
+                return odspSnapshot;
             } catch (error) {
                 const errorType = error.errorType;
                 if ((errorType === DriverErrorType.authorizationError || errorType === DriverErrorType.incorrectServerResponse) && tokenFetchOptions.refresh === false) {
                     throw error;
                 }
                 this.logger.sendErrorEvent({ eventName: "TreeLatest_FallBackToGetRequest" }, error);
-                return this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, false);
+                return this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, false);
             }
         } else {
-            return this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, false);
+            return this.fetchSnapshotWithTimeout(snapshotOptions, tokenFetchOptions, false);
+        }
+    }
+
+    private async fetchSnapshotWithTimeout(snapshotOptions: ISnapshotOptions, tokenFetchOptions: TokenFetchOptions, usePost: boolean) {
+        assert(snapshotOptions.timeout !== undefined, "Timeout should be provided for setting a limit");
+        const abortController = new AbortController();
+        setTimeout(
+            () => {
+                abortController.abort();
+            },
+            snapshotOptions.timeout,
+        );
+        try {
+            const odspSnapshot = await this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions, usePost, abortController);
+            return odspSnapshot;
+        } catch (error) {
+            if (error.errorType === OdspErrorType.snapshotTooBig || error.errorType === OdspErrorType.fetchTimeout) {
+                snapshotOptions.blobs = undefined;
```

Comment:
ahhh. here it is. should we shallow clone before changing the options, as this will change the object, and modify the behavior of the retry above, as this is nested

Reply:
should timeout and size be nulled out, what if the spine is HUGE

---
In reply to: [509817789](https://github.com/microsoft/FluidFramework/pull/4026#discussion_r509817789) [](ancestors = 509817789)

Reply:
my worry is we have a block somewhere over 25MB, especially once we have blob support. this code can never download it, but maybe that is ok

---
In reply to: [509819764](https://github.com/microsoft/FluidFramework/pull/4026#discussion_r509819764) [](ancestors = 509819764,509817789)

Reply:
Did the cloning. Also these limits will be tuned based on telemetry. So we can change this in future.